token/link/link.go

```diff
@@ -0,0 +1,217 @@
+	linkURL, err := url.Parse(redirectURL)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"redirect_url": redirectURL,
+			"err":          err,
+		}, "unable to parse redirectURL")
+		return "", errs.NewBadParameterError("redirect", redirectURL).Expected("valid URL")
+	}
+	parameters := linkURL.Query()
+	parameters.Add(identityIDParam, identityID)
+	parameters.Add(forParam, forResource)
+	linkURL.RawQuery = parameters.Encode()
+	redirectURL = linkURL.String()
+
+	oauthProvider, err := service.providerFactory.NewOauthProvider(ctx, req, service.config, forResource)
+	if err != nil {
+		return "", err
+	}
+	stateID := uuid.NewV4()
+	err = oauth.SaveReferrer(ctx, service.db, stateID, redirectURL, service.config.GetValidRedirectURLs())
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"redirect_url": redirectURL,
+			"err":          err,
+		}, "unable to save the state")
+		return "", err
+	}
+
+	return oauthProvider.AuthCodeURL(stateID.String(), oauth2.AccessTypeOnline), nil
+}
+
+// Callback returns a redirect URL after callback from an external oauth2 resource provider such as GitHub during user's account linking
+func (service *LinkService) Callback(ctx context.Context, req *goa.RequestData, state string, code string) (string, error) {
+	// validate known state
+	knownReferrer, err := oauth.LoadReferrer(ctx, service.db, state)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state": state,
+			"err":   err,
+		}, "can't load referrer by state")
+		return "", err
+	}
+
+	referrerURL, err := url.Parse(knownReferrer)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"code":           code,
+			"state":          state,
+			"known_referrer": knownReferrer,
+			"err":            err,
+		}, "failed to parse referrer")
+		return "", err
+	}
+
+	identityID := referrerURL.Query().Get(identityIDParam)
+	identityUUID, err := uuid.FromString(identityID)
+	if err != nil {
+		return "", err
+	}
+
+	forResource := referrerURL.Query().Get(forParam)
+
+	oauthProvider, err := service.providerFactory.NewOauthProvider(ctx, req, service.config, forResource)
+	if err != nil {
+		return "", err
+	}
+
+	providerToken, err := oauthProvider.Exchange(ctx, code)
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"state": state,
+			"code":  code,
+			"err":   err,
+		}, "exchange operation failed")
+		return "", err
+	}
+	if providerToken.AccessToken == "" {
+		log.Error(ctx, map[string]interface{}{
+			"state":       state,
+			"code":        code,
+			"provider_id": oauthProvider.ID(),
+		}, "access token return by provider is empty")
+		return "", errors.New("access token return by provider is empty")
+	}
+
+	err = application.Transactional(service.db, func(appl application.Application) error {
+		tokens, err := appl.ExternalTokens().LoadByProviderIDAndIdentityID(ctx, oauthProvider.ID(), identityUUID)
```

Comment:
Thinking we should check with "ProviderID + IdentityID + Scope" ? The scope will be default anyway for now. Let me update the repository API for that since this is the most popular use case "Get token for XXX provider for YYY identity for the scope ZZZ".

And if some token in the same scope was already present, we should overwrite it - else `create` a new entity in the table.

Reply:
The scope thing is more complicated actually. For example git has limitation of number of tokens per user&app. So, if we just blindly add new tokens with new scopes we may invalidate the existing tokens. Another concern is UX around it. It would be a bad UX if user just linked GitHub and granted a write access to OSIO (let's say in new app wizard) and we ask the same user later for read access when another client (Che for example) asked for it.
Token scopes will be in DB so we could utilize it later but for now I wouldn't do any logic around it.

Reply:
> For example git has limitation of number of tokens per user&app. So, if we just blindly add new tokens with new scopes we may invalidate the existing tokens

haha yeah, this actually puts the whole per-scope-per-token plan in jeopardy :(